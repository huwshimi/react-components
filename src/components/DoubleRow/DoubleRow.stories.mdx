import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import DoubleRow from "./DoubleRow";
import ModularTable from "../ModularTable";
import Icon, { ICONS } from "../Icon";

<Meta
  title="DoubleRow"
  component={DoubleRow}
  argTypes={{
    children: {
      control: {
        type: "text",
      },
    },
  }}
/>

export const Template = (args) => <DoubleRow {...args} />;

### DoubleRow

This is a double content row for use in tables.

### Props

<ArgsTable of={DoubleRow} />

### DoubleRow

<Canvas>
  <Story
    name="Default"
    args={{ primary: "eth0", secondary: "0d:be:38:1d:1c:a9" }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

See [Col](/#/Col) for grid details.

### Props

<ArgsTable of={DoubleRow} />

### With ModularTable

<Preview>
  <Story name="With ModularTable">
    <ModularTable
      columns={React.useMemo(
        () => [
          {
            Header: "ID",
            accessor: "build",
            Cell: ({ value }) => (
              <DoubleRow
                iconSpace
                icon={
                  (value.result === "released" && (
                    <Icon name={ICONS.success} />
                  )) ||
                  (value.result === "failed" && (
                    <Icon name={ICONS.error} />
                  )) || <Icon name="" />
                }
                primary={value.buildId}
                secondary={value.duration}
              />
            ),
          },
        ],
        []
      )}
      data={React.useMemo(
        () => [
          {
            build: {
              buildId: "5432",
              arch: "arm64",
              duration: "5 minutes",
              result: "released",
              finished: "10 minutes ago",
            },
          },
          {
            build: {
              buildId: "1234",
              arch: "armhf",
              duration: "5 minutes",
              result: "failed",
              finished: "over 1 year ago",
            },
          },
          {
            build: {
              buildId: "1111",
              arch: "test64",
              duration: "1 minute",
              result: "other",
              finished: "ages ago",
            },
          },
        ],
        []
      )}
    />
  </Story>
</Preview>
