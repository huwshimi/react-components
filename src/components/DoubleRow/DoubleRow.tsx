import { useRef } from "react";
import type { ReactNode } from "react";
import classNames from "classnames";

import TableMenu from "../TableMenu";
import type { Props as TableMenuProps } from "../TableMenu/TableMenu";

type Props<L> = {
  className?: string | null;
  icon?: ReactNode | null;
  iconSpace?: boolean | null;
  menuClassName?: string | null;
  menuLinks?: TableMenuProps<L>["links"] | null;
  menuTitle?: string | null;
  onToggleMenu?: TableMenuProps<L>["onToggleMenu"] | null;
  primary?: ReactNode | null;
  primaryAriaLabel?: string | null;
  primaryClassName?: string | null;
  primaryTextClassName?: string | null;
  primaryTitle?: string | null;
  secondary?: ReactNode | null;
  secondaryAriaLabel?: string | null;
  secondaryClassName?: string | null;
  secondaryTitle?: string | null;
};

const DoubleRow = <L,>({
  className,
  icon,
  iconSpace,
  menuClassName,
  menuLinks,
  menuTitle,
  onToggleMenu,
  primary,
  primaryAriaLabel,
  primaryClassName,
  primaryTextClassName,
  primaryTitle,
  secondary,
  secondaryAriaLabel,
  secondaryClassName,
  secondaryTitle,
}: Props<L>): JSX.Element => {
  const parent = useRef(null);
  const hasIcon = icon || iconSpace;

  return (
    <div
      className={classNames(
        {
          "p-double-row": !hasIcon,
          "p-double-row--with-icon": hasIcon,
        },
        className
      )}
    >
      {hasIcon ? (
        <div className="p-double-row__icon">
          {icon || <div className="p-double-row__icon-space"></div>}
        </div>
      ) : null}
      <div className="p-double-row__rows-container">
        <div
          aria-label={primaryAriaLabel}
          className={primaryClassName}
          ref={parent}
        >
          <div
            className={classNames("u-truncate", primaryTextClassName)}
            title={primaryTitle}
          >
            {primary}
          </div>
          {menuLinks ? (
            <TableMenu
              className={menuClassName}
              links={menuLinks}
              onToggleMenu={onToggleMenu}
              positionNode={parent.current}
              title={menuTitle}
            />
          ) : null}
        </div>
        {secondary ? (
          <div
            aria-label={secondaryAriaLabel}
            className={classNames("u-truncate", secondaryClassName)}
            data-test="secondary-row"
            title={secondaryTitle}
          >
            {secondary}
          </div>
        ) : null}
      </div>
    </div>
  );
};

export default DoubleRow;
