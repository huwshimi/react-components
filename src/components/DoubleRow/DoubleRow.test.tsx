import { shallow } from "enzyme";

import DoubleRow from "./DoubleRow";

describe("DoubleRow ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <DoubleRow primary="Top row" secondary="Bottom row" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can render without a secondary row", () => {
    const wrapper = shallow(<DoubleRow primary="Top row" />);
    expect(wrapper.find("[data-test='secondary-row']").exists()).toBe(false);
  });

  it("can display an icon", () => {
    const wrapper = shallow(
      <DoubleRow
        icon={<i className="p-icon"></i>}
        primary="Top row"
        secondary="Bottom row"
      />
    );
    expect(wrapper.find(".p-double-row--with-icon").exists()).toBe(true);
    expect(wrapper.find(".p-double-row__icon").exists()).toBe(true);
  });

  it("can display the space for an icon", () => {
    const wrapper = shallow(
      <DoubleRow iconSpace={true} primary="Top row" secondary="Bottom row" />
    );
    expect(wrapper.find(".p-double-row--with-icon").exists()).toBe(true);
    expect(wrapper.find(".p-double-row__icon").exists()).toBe(true);
  });

  it("can have a menu", () => {
    const wrapper = shallow(
      <DoubleRow
        menuLinks={[{ children: "Link1" }, { children: "Link2" }]}
        menuTitle="Take action:"
        primary="Top row"
        secondary="Bottom row"
      />
    );
    expect(wrapper.find("TableMenu").exists()).toBe(true);
  });
});
