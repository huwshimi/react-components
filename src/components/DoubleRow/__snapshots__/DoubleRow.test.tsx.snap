// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoubleRow  renders 1`] = `
<div
  className="p-double-row"
>
  <div
    className="p-double-row__rows-container"
  >
    <div
      className="p-double-row__primary-row"
    >
      <div
        className="p-double-row__primary-row-text u-truncate"
      >
        Top row
      </div>
    </div>
    <div
      className="p-double-row__secondary-row u-truncate"
    >
      Bottom row
    </div>
  </div>
</div>
`;
